src/agent/onefuzz-supervisor/src/heartbeat.rs

```diff
@@ -54,3 +53,28 @@ pub async fn init_agent_heartbeat(queue_url: Url) -> Result<AgentHeartbeatClient
     );
     Ok(hb)
 }
+
+pub trait HeartbeatSender {
+    fn send(&self, data: HeartbeatData) -> Result<()>;
+
+    fn alive(&self) {
+        self.send(HeartbeatData::MachineAlive).unwrap()
+    }
+}
+
+impl HeartbeatSender for AgentHeartbeatClient {
+    fn send(&self, data: HeartbeatData) -> Result<()> {
+        let mut messages_lock = self.context.pending_messages.lock().unwrap();
```

Comment:
How about a `?` here, instead, since you're already returning a `Result`?

Reply:
You could also permit failure, and maybe only return an error if lock acquisition fails `N` times, where `N` is compared against a counter value stored in `self`/`AgentHeartbeatClient`.

Reply:
? does not work here  i get this error
```
 --> onefuzz-supervisor\src\heartbeat.rs:67:69
   |
67 |         let mut messages_lock = self.context.pending_messages.lock()?;
   |                                                                     ^ `std::sync::MutexGuard<'_, std::collections::HashSet<heartbeat::HeartbeatData>>` cannot be sent between threads safely
   |
   = help: within `std::sync::PoisonError<std::sync::MutexGuard<'_, std::collections::HashSet<heartbeat::HeartbeatData>>>`, the trait `std::marker::Send` is not implemented for `std::sync::MutexGuard<'_, std::collections::HashSet<heartbeat::HeartbeatData>>`
   = note: required because it appears within the type `std::sync::PoisonError<std::sync::MutexGuard<'_, std::collections::HashSet<heartbeat::HeartbeatData>>>`
   = note: required because of the requirements on the impl of `std::convert::From<std::sync::PoisonError<std::sync::MutexGuard<'_, std::collections::HashSet<heartbeat::HeartbeatData>>>>` for `anyhow::Error`
   = note: required by `std::convert::From::from`
```
I removed the result return type. the expected behavior is to wait until the lock is free so no need to return a result.

Reply:
You could use map_err to convert the error to something that is `Send` here.